Decode special PMT readout channels

PetrilloAtWork opened the pull request:
This changes in the PMT decoder allow the decoding of the 24 PMT readout channels that are not associated to any channel ID in the PMT channel mapping database.
Each special channel must be explicitly configured in the FHiCL configuration in the module in order to be decoded.
The special channels can be written only if on global trigger time, only if there is signal on them, or always, and they can be redirected to a different data product (instance name) than the one for the "regular" channels.

The configuration provided as default for "Run1" includes:
* all the standard channels, as usual, in the data product `daqPMT` with no instance name
* 8 waveforms per trigger primitive, into the `daqPMT:trgprim` data product, with the digitization of the trigger primitive signal as entering the first of the three digitizers in a VME crate
* 2 waveforms featuring the BES signal, only when matching global trigger time, into the `daq:BES` data product
* all other 14 special channels, only for waveforms spanning at least 40 ADC, into the `daqPMT:unknwn` data product

I have successfully tested this on a data file of run `7863` using `decodePMT_icarus.fcl` file as test.
The test with `stage0_multiTPC_splitstream_nofilter_icarus.fcl` _failed_ for reasons unrelated (?) to this request, about CRT decoding.

Note that in the test with run `7863` no waveform landed in the `daqPMT:unknwn` data product.

Courtesy of `gh`:
- Removed use of obsolete utility util::fhicl::getOptionalValue()
- PMT decoding stores "special" channels

PetrilloAtWork commented:
Ideally, a reviewer would:
1. run a test with a recent raw data file and check that the PMT information that comes out is properly organised;
2. check the configuration file `CAEN_V1730_setup_icarus.fcl` for errors or things to set differently.

mmrosenberg commented:
trigger build

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/14853&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/14854&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/3933&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/3933&platform=Linux%203.10.0-1160.49.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14854&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for c7:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/3934&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/3934&platform=Linux%203.10.0-1160.49.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/14853&builds=)

SFBayLaser commented:
I am not sure I am expert enough on the details of the PMT output to satisfy the description of a reviewer given above. I think I will rely on Andrea's assessment and then sign off.

ascarpel reviewed (changes requested):
Only the names assigned to the signals sampled in the special channels needs to be changed. The schema is the following: 


I notice here a small mistake that I missed before. The signals sent to each spare channel of the crates are respectively:
    First digitizer: trigger primitive (correct)
    Second digitizer: BNB beam signal ( GatedBES or RWM, whatever is found better )
    Third digitizer: NuMI beam signal ( MIBS$74 or RWM, whatever is found better )

PetrilloAtWork commented:
After the first @ascarpel comment and discussion with him on a separate channel, I have changed the configuration in the following way:
* the 16th channel of the first board in each crate is still the trigger primitive, as before
* the 16th channel of the second board in each crate is the BNB gate signal, directed into `BNB` instance name
* the 16th channel of the third board in each crate is the NuMI gate signal, directed into `NuMI` instance name

The last two groups of waveforms are saved only at global trigger time, and only if there is signal (peak-to-peak larger than 100 ADC, that currently is ~12 mV).

I have added a configuration check that I realised was missing and easy to implement, too.

PetrilloAtWork commented on icaruscode/Decode/fcl/CAEN_V1730_setup_icarus.fcl:
```diff
@@ -109,10 +109,148 @@ icarus_V1730_setup_Run0: [
+  #
+  # top
+  #
+  { Name: "icaruspmtwetop01"  TriggerDelay: " 0 ns"  SpecialChannels: [ { @table::icarus_V1730_16thChannel_trgprim_setup_Run1  Channel: 0x1051 } ] },
```
I argue that since the signals are affected by those delays, they should be corrected natively here.
This allows to compare e.g. the BNB beam signal from the second readout board to the trigger primitive signal from the first readout board.

rennney commented:
trigger build

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/15026&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/15025&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for c7:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/4053&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/4053&platform=Linux%203.10.0-1160.49.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/15025&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/4052&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/4052&platform=Linux%203.10.0-1160.49.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/15026&builds=)

SFBayLaser reviewed (approved):
Ok, seems fine now
